support foreign threads, and adding threads at run time

JeffBezanson opened the pull request:
Hook a couple functions (notably cfunction and jl_call*) to handle adding foreign threads automatically.

Fixes #17573

ianatol commented on src/partr.c:
```diff
@@ -26,6 +26,9 @@ static const int16_t not_sleeping = 0;
 // it is acceptable for the thread to be sleeping.
 static const int16_t sleeping = 1;
 
+// this thread is dead.
+static const int16_t sleeping_like_the_dead = 2;
+
```
```suggestion

```

Seems like this is **dead** code 😄 (triggered Werror)

ianatol replied:
```suggestion

```

Seems like this is **dead** code 😄 (triggered Werror)

vtjnash commented on base/threadingconstructs.jl:
```diff
@@ -57,10 +57,21 @@ Returns the number of threadpools currently configured.
+[`LinearAlgebra`](@ref man-linalg) standard library, and `nprocs()` in the
+[`Distributed`](@ref man-distributed) standard library.
+"""
+nthreadpool() = Threads._nthreads_in_pool(Int8(0))
```
Choose a better name, to be less confusable with `nthreadpools`. Possibly `sizethreadpool`?

tkf replied:
`defaultpoolsize`?

IanButterworth replied:
Why not
`nthreads()` -> default pool
`nthreads(pool)`

`nthreadpool` seems confusing IMO

vtjnash replied:
`nthreads()` means something different now, so that would probably violate semver

JeffBezanson replied:
True, `nthreads()` currently means the total number of threads over all pools, so we should just keep it that way. `nthreads(pool)` seems fine to me too. If we want a new function for the number of threads in the default pool we can maybe add that, but it doesn't need to be done in this PR.

vtjnash replied:
But after rebasing, this seems to be running into some sort of memory issue now, so I am trying to figure out what it is conflicting with

vtjnash commented on base/threads.jl:
```diff
@@ -11,25 +11,4 @@ include("threadingconstructs.jl")
-    end
-    return A
-end
-
```
Would a more object-oriented API be good for replacing this?
```julia
struct ThreadLocalData
    f::Function
    l::Lock
    d::Vector{Any}
end
Base.getindex(tld::ThreadLocalData) = ...
```

e.g. https://github.com/JuliaLang/julia/issues/14135

tkf replied:
ref https://juliaconcurrent.github.io/ConcurrentUtils.jl/dev/#ConcurrentUtils.ThreadLocalStorage

vtjnash committed:
[Profile] allocate exactly the space requested

With dynamic thread counts, we cannot ensure this count is constant
after initialization, and we might like to even profile adding and
removing threads.

tkf commented on base/threadingconstructs.jl:
```diff
@@ -11,21 +10,21 @@ ID `1`.
-pools is the inclusive upper bound on [`threadid()`](@ref).
+Get a lower bound on the number of threads (across all thread pools or within
+the specified thread pool) available to the Julia process >= [`threadid()`](@ref),
+with atomic-acquire semantics.
```
I find it strange that acquire ordering is mentioned without noting what events have release store. Also, this is not an accurate description of the method `maxthreadid(pool::Symbol)` since it uses non-atomic load (the first load is OK):

https://github.com/JuliaLang/julia/blob/d635803c50b3ce48fd4d9d74638fdc07b1dd6dd0/base/threadingconstructs.jl#L38-L41

Also, my impression is that this ordering is required for runtime authors and not Julia users. For example, monotonic load with `max(threadid(), maxthreadid())` is sufficient for supporting DCL thread-local storage initialization pattern.

I think it'd be simpler to not document the ordering in the public API specification.

tkf commented on base/threadingconstructs.jl:
```diff
@@ -11,21 +10,21 @@ ID `1`.
 threadid() = Int(ccall(:jl_threadid, Int16, ())+1)
 
+# lower bound on the largest threadid()
 """
-    Threads.nthreads([:default|:interactive]) -> Int
+    Threads.maxthreadid([:default|:interactive]) -> Int
```
How does `maxthreadid(pool::Symbol)` work when allowing adding threads to both worker pools? Are thread IDs for each pool non-contiguous now? Or do thread IDs of interactive pool threads change after adding a thread to the default pool?

vtjnash committed:
threading: support more than nthreads at runtime

Hook a couple functions (notably cfunction) to handle adopting
foreign threads automatically when used.

n.b. If returning an object pointer, we do not gc_unsafe_leave
afterwards as that would render the pointer invalid. However, this means
that it can be a long time before the next safepoint (if ever). We
should look into ways of improving this bad situation, such as pinning
only that specific object temporarily.

n.b. There are some remaining issues to clean up. For example, we may
trap pages in the ptls after GC to keep them "warm", and trap other
pages in the unwind buffer, etc.

JeffBezanson commented on base/threadingconstructs.jl:
```diff
@@ -11,21 +10,21 @@ ID `1`.
 
-nthreads() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
-function nthreads(pool::Symbol)
+function maxthreadid(pool::Symbol)
```
Looks like a merge conflict --- this could still be called `nthreads`.

IanButterworth commented on base/threadingconstructs.jl:
```diff
@@ -1,7 +1,6 @@
 
-export threadid, nthreads, @threads, @spawn,
-       threadpool, nthreadpools
+export threadid, @threads, @spawn, threadpool, nthreadpools
```
`nthreads` is deprecated but should still be exported, otherwise it would be a breaking change?
Also, will this work given that the deprecation is `@eval`-ed?
```suggestion
export threadid, nthreads, @threads, @spawn, threadpool, nthreadpools
```

vtjnash replied:
There is problems in [SparseArrays]( https://github.com/JuliaSparse/SparseArrays.jl/blob/2bbdd7a12ead8207593655c541ba347761a9c663/src/solvers/cholmod.jl#L224), so I couldn't make it a proper deprecation yet until that package is fixed

JeffBezanson replied:
We need to put this back until we can do a proper deprecation though.

IanButterworth replied:
Re. The SparseArrays issue https://github.com/JuliaSparse/SparseArrays.jl/pull/204

Wimmerer replied:
See https://github.com/JuliaSparse/SparseArrays.jl/pull/206

IanButterworth replied:
SparseArrays is now fixed on julia master

DilumAluthge commented:
Needs a PkgEval run to evaluate the impact of deleting the `Base.Threads.nthreads` function.

DilumAluthge commented:
@nanosoldier `runtests(ALL, vs = ":master")`

maleadt commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45447#issuecomment-1205770856) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8a68504_vs_de375ef/report.html).

vtjnash commented:
Looks like there are a couple more key packages that need fixing from glancing at some parts of that report, including FFTW, HTTP, and Parsers. The HTTP and Parsers code looks like it was probably already very dangerous (looks like it probably may be introducing data races into the code). The FFTW code is just trivially deciding if thread support is enabled, and that call can be removed: https://github.com/JuliaMath/FFTW.jl/blob/17bc81a0fcf9875d777ea4bee2fca70fc23c8a0c/src/providers.jl#L75

DilumAluthge commented:
Additionally, the PR fails to build on `i686` and fails tests on several platforms.

IanButterworth commented on src/signals-unix.c:
```diff
@@ -701,7 +708,7 @@ void trigger_profile_peek(void)
     if (bt_size_max == 0){
         // If the buffer hasn't been initialized, initialize with default size
         // Keep these values synchronized with Profile.default_init()
-        if (jl_profile_init(10000000 * jl_n_threads, 1000000) == -1){
+        if (jl_profile_init(10000000, 1000000) == -1) {
```
There's a note in the `Profile.init()` docstring that needs updating if this is being reverted. Also the revert should probably be backported to 1.8 otherwise it's the only minor that's going to have that feature
